Pull request for Raw Value in DiscoverySelectors

bechte opened the pull request:
## Overview

This pull requests solves the following issues:

- Store raw values in DiscoverySelectors & TestSources #489
- Cannot select default method by fully qualified method name #516
- Non-standard method names are not supported for discovery selection #477

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

jbduncan commented on junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolver.java:
```diff
@@ -24,11 +24,8 @@
-import org.junit.platform.engine.discovery.JavaMethodSelector;
-import org.junit.platform.engine.discovery.JavaPackageSelector;
-import org.junit.platform.engine.discovery.UniqueIdSelector;
+import org.junit.platform.engine.discovery.*;
```
I think this would be clearer as a bunch of explicit imports, rather than a wildcard.

bechte replied:
fine with that one.

jbduncan commented on junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/discovery/DiscoverySelectorResolverTests.java:
```diff
@@ -50,11 +50,8 @@
-import org.junit.platform.engine.discovery.JavaMethodSelector;
-import org.junit.platform.engine.discovery.JavaPackageSelector;
-import org.junit.platform.engine.discovery.UniqueIdSelector;
+import org.junit.platform.engine.discovery.*;
```
Could this be a set of explicit imports as well?

bechte replied:
I would leave this one, as it is in a test class that uses mostly all.

jbduncan commented on platform-tests/src/test/java/org/junit/platform/engine/discovery/DiscoverySelectorsTests.java:
```diff
@@ -14,20 +14,13 @@
-import static org.junit.platform.engine.discovery.DiscoverySelectors.selectJavaMethod;
-import static org.junit.platform.engine.discovery.DiscoverySelectors.selectJavaPackage;
-import static org.junit.platform.engine.discovery.DiscoverySelectors.selectUri;
+import static org.junit.platform.engine.discovery.DiscoverySelectors.*;
```
Could this be done as a set of explicit imports as well?

bechte replied:
see above

jbduncan commented on platform-tests/src/test/java/org/junit/platform/launcher/core/LauncherDiscoveryRequestBuilderTests.java:
```diff
@@ -40,11 +40,7 @@
-import org.junit.platform.engine.discovery.JavaMethodSelector;
-import org.junit.platform.engine.discovery.JavaPackageSelector;
-import org.junit.platform.engine.discovery.UniqueIdSelector;
+import org.junit.platform.engine.discovery.*;
```
What about this line as well, regarding explicit imports?

bechte replied:
see above

marcphilipp commented on junit-vintage-engine/src/test/java/org/junit/vintage/engine/VintageTestEngineDiscoveryTests.java:
```diff
@@ -676,9 +677,9 @@ private static void assertClassSource(Class<?> expectedClass, TestDescriptor tes
 	private static void assertMethodSource(Method expectedMethod, TestDescriptor testDescriptor) {
 		assertThat(testDescriptor.getSource()).containsInstanceOf(JavaMethodSource.class);
 		JavaMethodSource methodSource = (JavaMethodSource) testDescriptor.getSource().get();
-		assertThat(methodSource.getJavaClass()).isEqualTo(expectedMethod.getDeclaringClass());
```
There should be a check for `getClassName()` now, shouldn't there?

bechte replied:
True, missed that. Thanks. I will add it.

bechte commented:
I updated the branch with a `rebase` and a `push --force`. It seems that github will not detect changes this way. It does not resolve the change request by @marcphilipp even though I changed the code. Should we avoid to use `rebase & push --force` on PRs?

marcphilipp commented:
That's probably because I commented on a line that is still deleted. I've manually approved your changes now. Thanks!

mmerdes commented on junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/JavaClassSelector.java:
```diff
@@ -13,36 +13,60 @@
+		if (this.javaClass == null) {
+			this.javaClass = ReflectionUtils.loadClass(this.className).orElseThrow(
+				() -> new PreconditionViolationException("Could not load class with name: " + this.className));
+		}
```
Any danger we might run into thread safety problems here?

bechte replied:
Do you mean race conditions? It might but it's not expected that two threads will call it. And if so, it's lazy loading resulting in the same state. So no troubles if two threads calculate the result.

mmerdes commented on platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/JavaSourceTests.java:
```diff
@@ -59,6 +59,26 @@ void packageSourceFromPackageReference() {
 
 	@Test
+	void classNameSource() {
+		String testClassName = "com.unknown.package.ClassByName";
```
Maybe we should avoid packages called 'package'?
:)

bechte replied:
I changed it to `mypackage`. ;)

mmerdes commented on junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java:
```diff
@@ -282,20 +287,24 @@ public static JavaClassSelector selectJavaClass(String className) {
+		Preconditions.condition(matcher.matches(), "Method name is not a valid full qualified method name");
 
-		return selectJavaMethod(method.getDeclaringClass(), method);
+		if (StringUtils.isNotBlank(matcher.group(3))) {
```
I would prefer to have the various `matcher.group(index)`-calls assigned to variables with explanatory names for better readability

bechte replied:
I updated the PR.

sbrannen commented:
Hi guys,

Regarding wildcards for imports, we actually agreed on using implicit imports (for both non-static and static imports) for up to 20 imports from the same package/class and only using `*` after 20.

I guess this may not actually be documented or enforced anywhere, _but_ we did agree on that at some point. 😉

bechte commented:
According imports:
I've changed my IDE to exactly the same numbers (20 for both, static and non-static) to avoid auto-replacements from my side.

According `javaClass`:
Running on the JVM, I dont feel so bad calling it `javaClass`, actually. ;-) It's personal taste (just like in many cases with naming), but I dont like misspelling things, e.g. clazz, paggage, etc. just because of restrictions. I rather have a prefix / suffix. I will rename the `javaClass` back to `clazz` during the renaming of `JavaClassSelector` to `ClassSelector` if that is fine with you.
